src/js/binaryen.js-post.js

```diff
@@ -1185,16 +1185,42 @@ Module['Module'] = function(module) {
     Module['_BinaryenModuleDispose'](module);
   };
   var MAX = 1024*1024; // TODO: fix this hard-wired limit
-  var writeBuffer = null;
-  this['emitBinary'] = function() {
-    if (!writeBuffer) writeBuffer = _malloc(MAX);
-    var bytes = Module['_BinaryenModuleWrite'](module, writeBuffer, MAX);
+  var outputBuffer = null;
+  var sourceMapBuffer = null;
+  this['emitBinary'] = function(sourceMapUrl) {
+    if (!outputBuffer) outputBuffer = _malloc(MAX);
+    var bytes = Module['_BinaryenModuleWrite'](module, outputBuffer, MAX);
     assert(bytes < MAX, 'FIXME: hardcoded limit on module size'); // we should not use the whole buffer
-    return new Uint8Array(HEAPU8.subarray(writeBuffer, writeBuffer + bytes));
+    return new Uint8Array(HEAPU8.subarray(outputBuffer, outputBuffer + bytes));
+  };
+  this['emitBinaryWithSourceMap'] = function(sourceMapUrl) {
+    if (!outputBuffer) outputBuffer = _malloc(MAX);
+    if (!sourceMapBuffer) sourceMapBuffer = _malloc(MAX);
+    return preserveStack(function() {
+      Module['_BinaryenModuleWriteWithSourceMap'](temp, module, strToStack(sourceMapUrl), outputBuffer, MAX, sourceMapBuffer, MAX);
```

Comment:
(looks like this depends on the ABI issues mentioned above, it assumes the output is pointed to by an extra param at the beginning? I'm actually surprised it's not at the end.)

Reply:
I just reused what's done for `BinaryenLiteral`, which is also a returned struct, above, and reused its temporary buffer. I actually don't know how exactly this works :)

Reply:
Oh, I see. Yeah, we already do depend on those ABI details I guess... no big deal to add this, then.